lib/foreman/logging.rb

```diff
@@ -0,0 +1,159 @@
+      @config = YAML.load_file('config/logging.yaml')
+      @config = @config[:default].deep_merge(@config[environment.to_sym])
+    end
+
+    def ensure_log_directory(log_directory)
+      return true if File.directory?(log_directory)
+
+      begin
+        FileUtils.mkdir_p(log_directory)
+      rescue Errno::EACCES
+        warn "Insufficient privileges for #{log_directory}"
+      end
+    end
+
+    # set root logger specific @config
+    #
+    # root logger has configurable appender by argument +options+
+    # we also set fallback appender to STDOUT in case a developer asks for unusable appender
+    def configure_root_logger(options)
+      ::Logging.logger.root.level     = @config[:level]
+      ::Logging.logger.root.appenders = build_root_appender(options)
+      ::Logging.logger.root.trace     = @config[:log_trace]
+
+      # fallback to log to STDOUT if there is any @config problem
+      if ::Logging.logger.root.appenders.empty?
+        ::Logging.logger.root.appenders = ::Logging.appenders.stdout
+        ::Logging.logger.root.warn 'No appender set, logging to STDOUT'
+      end
+    end
+
+    def build_console_appender
+      return unless @config[:console_inline]
+
+      ::Logging.logger.root.add_appenders(
+        ::Logging.appenders.stdout(:layout => build_layout(@config[:pattern], @config[:colorize]))
+      )
+    end
+
+    # currently we support two types of appenders, rolling file and syslog
+    # note that syslog ignores pattern and logs only messages
+    def build_root_appender(options)
+      name = "foreman"
+
+      case @config[:type]
+      when 'syslog'
+        build_syslog_appender(name, options)
+      when 'file'
+        build_file_appender(name, options)
+      else
+        fail 'unsupported logger type, please choose syslog or file'
+      end
+    end
+
+    def build_syslog_appender(name, options)
+      ::Logging.appenders.syslog(name, options.reverse_merge(:facility => ::Syslog::Constants::LOG_DAEMON))
+    end
+
+    def build_file_appender(name, options)
+      log_filename = "#{@log_directory}/#{@config[:filename]}"
+      begin
+        ::Logging.appenders.file(
+          name,
+          options.reverse_merge(:filename => log_filename,
+                                :layout   => build_layout(@config[:pattern], @config[:colorize]))
+        )
+      rescue ArgumentError
+        # if appender cannot open output file we ignore it, STDOUT fallback will be used
+        nil
+      end
+    end
+
+    def build_layout(pattern, colorize)
+      pattern += "  Log trace: %F:%L method: %M\n" if @config[:log_trace]
+      MultilinePatternLayout.new(:pattern => pattern, :color_scheme => colorize ? 'bright' : nil)
+    end
+
+    def configure_color_scheme
+      ::Logging.color_scheme(
+        'bright',
+        :levels => {
+          :info  => :green,
+          :warn  => :yellow,
+          :error => :red,
+          :fatal => [:white, :on_red]
+        },
+        :date   => :blue,
```

Comment:
At the risk of starting a bikeshed discussion, this is rather unreadable for me on a black terminal.